server/src/test/java/org/elasticsearch/cluster/metadata/MetadataIndexStateServiceBatchingTests.java

```diff
@@ -0,0 +1,207 @@
+        assertBusy(() -> assertThat(masterService.pendingTasks(), hasSize(3))); // two plus the blocking task itself
+
+        block2.call(); // release block
+
+        // assert that the requests were acknowledged
+        final var resp1 = future1.get();
+        assertAcked(resp1);
+        assertThat(resp1.getIndices(), hasSize(1));
+        assertThat(resp1.getIndices().get(0).getIndex().getName(), is("test-1"));
+
+        final var resp2 = future2.get();
+        assertAcked(resp2);
+        assertThat(resp2.getIndices(), hasSize(2));
+        assertThat(resp2.getIndices().stream().map(r -> r.getIndex().getName()).toList(), containsInAnyOrder("test-2", "test-3"));
+    }
+
+    public void testBatchBlockIndices() throws Exception {
+        final var masterService = getInstanceFromNode(ClusterService.class).getMasterService();
+
+        // create some indices
+        createIndex("test-1", client().admin().indices().prepareCreate("test-1"));
+        createIndex("test-2", client().admin().indices().prepareCreate("test-2"));
+        createIndex("test-3", client().admin().indices().prepareCreate("test-3"));
+        ensureGreen("test-1", "test-2", "test-3");
+
+        final var block1 = blockMasterService(masterService);
+        block1.call(); // wait for block
+
+        // fire off some closes
+        final var future1 = client().admin().indices().prepareAddBlock(APIBlock.WRITE, "test-1").execute();
+        final var future2 = client().admin().indices().prepareAddBlock(APIBlock.WRITE, "test-2", "test-3").execute();
+
+        // check the queue for the first add-block tasks (the add-index-block tasks)
+        assertBusy(() -> assertThat(masterService.pendingTasks(), hasSize(3))); // two plus the blocking task itself
+
+        // add *another* block to the end of the pending tasks, then unblock the current block so we can progress,
+        // then immediately block again on that new block
+        final var block2 = blockMasterService(masterService);
+        block1.call(); // release block
+        block2.call(); // wait for block
+
+        // check the queue for the second add-block tasks (the finalize-index-block tasks)
+        assertBusy(() -> assertThat(masterService.pendingTasks(), hasSize(3))); // two plus the blocking task itself
+
+        block2.call(); // release block
+
+        // assert that the requests were acknowledged
+        final var resp1 = future1.get();
+        assertAcked(resp1);
+        assertThat(resp1.getIndices(), hasSize(1));
+        assertThat(resp1.getIndices().get(0).getIndex().getName(), is("test-1"));
+
+        final var resp2 = future2.get();
+        assertAcked(resp2);
+        assertThat(resp2.getIndices(), hasSize(2));
+        assertThat(resp2.getIndices().stream().map(r -> r.getIndex().getName()).toList(), containsInAnyOrder("test-2", "test-3"));
+    }
+
+    private static Callable<Void> blockMasterService(MasterService masterService) {
```

Comment:
Why `Callable<Void>` and not just `CheckedRunnable<Exception>`?

Reply:
Why, because I didn't know that existed of course. 😆

Reply:
bb3839552d7c254e1b7b1cb395f6d49bb9cc0f72